### ISSUE TITLE
BUG in while loop (perhaps more) - see later comment on similar cases!

### ISSUE THREAD

**Issue by catmando** [opal/opal#761]
Just discovered this... looks like if you drop off the end of a method, and the last statement was a while, it returns javascript undefined.  

Here is a simple test that runs in opal-rspec

```
class Test

    def busted_while(x, y)
      while x < 10
        return true if x==y
        x += 1
      end
    end

  end

  describe "a busted while loop 1" do
    it "works if returning explicitly" do
      expect(Test.new.busted_while(7, 8)).to be_truthy
    end
    it "fails if there is no explicit return" do
      result = Test.new.busted_while(7, 5)
      expect(!result).to be_truthy
    end
  end
```

The second test bombs with "Cannot read property '$!' of undefined"

it you put any statement between the end of the while, and the end of the method all will be fine.

**Comment by vais** [comment]
:+1: @catmando for spotting this. @adambeynon I am trying to understand a little how the compiler works, but have no idea where to start figuring out this issue.

This:
```ruby
def test
  while false
  end
  'hi'
end
```
Compiles to this:
```javascript
/* Generated by Opal 0.7.0 */
(function(Opal) {
  Opal.dynamic_require_severity = "error";
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  return (Opal.Object.$$proto.$test = function() {
    var $a, $b, self = this;

    while ((($b = false) !== nil && (!$b.$$is_boolean || $b == true))) {
    };
    return "hi";
  }, nil) && 'test'
})(Opal);
```
But this:
```ruby
def test
  while false
  end
end
```
Compiles to this:
```javascript
/* Generated by Opal 0.7.0 */
(function(Opal) {
  Opal.dynamic_require_severity = "error";
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  return (Opal.Object.$$proto.$test = function() {
    var $a, $b, self = this;

    while ((($b = false) !== nil && (!$b.$$is_boolean || $b == true))) {
    };
  }, nil) && 'test'
})(Opal);
```
Note the absence of the `return` statement. Where in the compiler code does this get decided?

**Comment by adambeynon** [comment]
@vais The compiler has a returns method:

https://github.com/opal/opal/blob/master/lib/opal/compiler.rb#L302

This examines special case nodes which need to be handled in a certain way. For example, it makes sure an if-else expression returns its true or false case if it needs to. For some reason, a while expression has its special case commented out:

https://github.com/opal/opal/blob/master/lib/opal/compiler.rb#L346

This would ensure its value gets returned. Im not sure why it is commented out however, it must have been buggy at some point and never actually fixed. Its a bit of a mess the way it works, so I can take a look. A while expression might need some special extra work, Im not sure on the MRI semantics - does it return the last truth value, for example? Or always nil? (currently away from a laptop to test).

**Comment by vais** [comment]
@adambeynon `while` always evaluates to `nil`:
```
$ pry
[1] pry(main)> i = 0; while i < 1; i += 1; puts i; i; end
1
=> nil
```
But there is a catch (quoting from http://ruby-doc.org/core-2.1.1/doc/syntax/control_expressions_rdoc.html#label-while+Loop):
> The result of a while loop is nil unless break is used to supply a value.

So:
```
$ pry
[1] pry(main)> i = 0; while i < 1; break 'hi'; i += 1; puts i; i; end
=> "hi"
```

**Comment by vais** [comment]
@adambeynon BTW, simply un-commenting that line in `Compiler#returns` causes 289 spec failures, so, it's commented out for a good reason :)

**Comment by catmando** [comment]
I think there are more of these:  I don't have the time right now to turn this into a failure case but here is similar code snippet where adding a nil, before the end of the block is necessary to keep the caller from failing.  

```ruby
  def self.ready
    return unless block_given?
    if !@_promises 
      yield
    elsif @_promises.resolved?
      yield 
    else
      @_promises = @_promises.then { yield }
      nil  
      # ****** without this nil eventually the caller (or its caller, etc) will fail **********
      # NOTE at this point @_promises was ALWAYS initialized like this 
      # @_promises = Promise.when(...)  so perhaps Promise.when a while loop 
      # that is yielding to the then above???
    end
  end
```

**Comment by elia** [comment]
Should be fixed since at least 0.11

```rb
i = 0
a = while i <= 100
  break 'hi'
end
p a: a
```

```
⤑ pbpaste | ruby -v -                                                          ~
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin17]
{:a=>"hi"}
⤑ pbpaste | opal -v -                                                          ~
Opal v0.11.3
{"a"=>"hi"}
```
